### PR TITLE
Metacal catalog for dr1b

### PR DESCRIPTION
**Pull request by EiffL**
This PR adds the aggregated metacal catalog for dr1b, but I'm having a problem I can't figure out:
```
# Temporary fix while GCRCatalogs is updated
import sys
sys.path.insert(0, '/global/homes/f/flanusse/repo/gcr-catalogs/') # You need the master branch of gcr-catalogs: https://github.com/LSSTDESC/gcr-catalogs
import GCRCatalogs
from GCR import GCRQuery

catalog = GCRCatalogs.load_catalog('dc2_object_run2.1i_dr1b_with_metacal_griz.yaml');

res_ra = []
res_dec  = []
res_inds = []
for res in catalog.get_quantities(['ra','dec', 'id', 'mag_r_cModel', 'mag_g_cModel', 'extendedness' , 'blendedness', 'mcal_flags', 'tract', 'patch'], return_iterator=True):
    m = res['mcal_flags'] == 0
    res_ra.append(res['ra'][m])
    res_dec.append(res['dec'][m])
    res_inds.append(res['id'][m])
    print(unique(res['tract'][m]), unique(res['patch'][m]))
```

crashes for tract 2897 with following message:
```
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-3-0f08ad40ca1c> in <module>
      2 res_dec  = []
      3 res_inds = []
----> 4 for res in catalog.get_quantities(['ra','dec', 'id', 'mag_r_cModel', 'mag_g_cModel', 'extendedness' , 'blendedness', 'mcal_flags', 'tract', 'patch'], return_iterator=True):
      5     m = res['mcal_flags'] == 0
      6     res_ra.append(res['ra'][m])

~/.local/lib/python3.7/site-packages/GCR/base.py in _get_quantities_iter(self, quantities, filters, native_filters)
    460         for native_quantity_getter in self._iter_native_dataset(native_filters):
    461             data = self._load_quantities(quantities.union(set(filters.variable_names)),
--> 462                                          native_quantity_getter)
    463             data = filters.filter(data)
    464             for q in set(data).difference(quantities):

~/.local/lib/python3.7/site-packages/GCR/base.py in _load_quantities(self, quantities, native_quantity_getter)
    454     def _load_quantities(self, quantities, native_quantity_getter):
    455         native_quantities_needed = self._translate_quantities(quantities)
--> 456         native_data = self._obtain_native_data_dict(native_quantities_needed, native_quantity_getter)
    457         return {q: self._assemble_quantity(q, native_data) for q in quantities}
    458 

~/.local/lib/python3.7/site-packages/GCR/composite.py in _obtain_native_data_dict(self, native_quantities_needed, native_quantity_getter)
    196                 for q, v in catalog.instance._load_quantities( # pylint: disable=W0212
    197                         native_quantities_needed_dict[catalog.identifier],
--> 198                         native_quantity_getter[catalog.identifier],
    199                 ).items():
    200                     data[(catalog.identifier, q)] = v

~/.local/lib/python3.7/site-packages/GCR/base.py in _load_quantities(self, quantities, native_quantity_getter)
    454     def _load_quantities(self, quantities, native_quantity_getter):
    455         native_quantities_needed = self._translate_quantities(quantities)
--> 456         native_data = self._obtain_native_data_dict(native_quantities_needed, native_quantity_getter)
    457         return {q: self._assemble_quantity(q, native_data) for q in quantities}
    458 

~/.local/lib/python3.7/site-packages/GCR/base.py in _obtain_native_data_dict(native_quantities_needed, native_quantity_getter)
    450         ```
    451         """
--> 452         return {q: native_quantity_getter(q) for q in native_quantities_needed}
    453 
    454     def _load_quantities(self, quantities, native_quantity_getter):

~/.local/lib/python3.7/site-packages/GCR/base.py in <dictcomp>(.0)
    450         ```
    451         """
--> 452         return {q: native_quantity_getter(q) for q in native_quantities_needed}
    453 
    454     def _load_quantities(self, quantities, native_quantity_getter):

~/repo/gcr-catalogs/GCRCatalogs/dc2_object.py in __getitem__(self, key)
    129         """
    130         if self._cache is None:
--> 131             self._cache = self.storer.read()
    132 
    133         try:

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/io/pytables.py in read(self, start, stop, **kwargs)
   2984             _start, _stop = (start, stop) if i == select_axis else (None, None)
   2985             ax = self.read_index('axis{idx}'.format(
-> 2986                 idx=i), start=_start, stop=_stop)
   2987             axes.append(ax)
   2988 

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/io/pytables.py in read_index(self, key, **kwargs)
   2573             return self.read_sparse_intindex(key, **kwargs)
   2574         elif variety == u'regular':
-> 2575             _, index = self.read_index_node(getattr(self.group, key), **kwargs)
   2576             return index
   2577         else:  # pragma: no cover

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/io/pytables.py in read_index_node(self, node, start, stop)
   2707             index = factory(_unconvert_index(data, kind,
   2708                                              encoding=self.encoding,
-> 2709                                              errors=self.errors), **kwargs)
   2710 
   2711         index.name = name

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/io/pytables.py in _unconvert_index(data, kind, encoding, errors)
   4635     elif kind in (u'string'):
   4636         index = _unconvert_string_array(data, nan_rep=None, encoding=encoding,
-> 4637                                         errors=errors)
   4638     elif kind == u'object':
   4639         index = np.asarray(data[0])

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/io/pytables.py in _unconvert_string_array(data, nan_rep, encoding, errors)
   4721 
   4722         if isinstance(data[0], compat.binary_type):
-> 4723             data = Series(data).str.decode(encoding, errors=errors).values
   4724         else:
   4725             data = data.astype(dtype, copy=False).astype(object, copy=False)

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/core/strings.py in decode(self, encoding, errors)
   2650     @copy(str_decode)
   2651     def decode(self, encoding, errors="strict"):
-> 2652         result = str_decode(self._parent, encoding, errors)
   2653         return self._wrap_result(result)
   2654 

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/core/strings.py in str_decode(arr, encoding, errors)
   1708         decoder = codecs.getdecoder(encoding)
   1709         f = lambda x: decoder(x, errors)[0]
-> 1710     return _na_map(f, arr)
   1711 
   1712 

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/core/strings.py in _na_map(f, arr, na_result, dtype)
     57 def _na_map(f, arr, na_result=np.nan, dtype=object):
     58     # should really _check_ for NA
---> 59     return _map(f, arr, na_mask=True, na_value=na_result, dtype=dtype)
     60 
     61 

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/core/strings.py in _map(f, arr, na_mask, na_value, dtype)
     72         try:
     73             convert = not all(mask)
---> 74             result = lib.map_infer_mask(arr, f, mask.view(np.uint8), convert)
     75         except (TypeError, AttributeError) as e:
     76             # Reraise the exception if callable `f` got wrong number of args.

pandas/_libs/lib.pyx in pandas._libs.lib.map_infer_mask()

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/site-packages/pandas/core/strings.py in <lambda>(x)
   1707     else:
   1708         decoder = codecs.getdecoder(encoding)
-> 1709         f = lambda x: decoder(x, errors)[0]
   1710     return _na_map(f, arr)
   1711 

/cvmfs/sw.lsst.eu/linux-x86_64/lsst_distrib/w_2019_19/python/miniconda3-4.5.12/envs/lsst-scipipe-1172c30/lib/python3.7/encodings/utf_8.py in decode(input, errors)
     14 
     15 def decode(input, errors='strict'):
---> 16     return codecs.utf_8_decode(input, errors, True)
     17 
     18 class IncrementalEncoder(codecs.IncrementalEncoder):

UnicodeDecodeError: 'utf-8' codec can't decode byte 0x83 in position 12: invalid start byte
```

Yet, no error when I load only the smaller 'dc2_object_run2.1i_dr1b_tract2897_with_metacal_griz.yaml' catalog

### CODE REVIEW

**Comment by EiffL** [comment]
@yymao @wmwv any thoughts?

File: GCRCatalogs/catalog_configs/dc2_metacal_griz_run2.1i_dr1b.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+subclass_name: dc2_metacal.DC2MetacalCatalog
+base_dir: /global/homes/f/flanusse/scratch/object_mcal_catalog_coaddv1_dr1b
```
Comment (yymao) [inline GCRCatalogs/catalog_configs/dc2_metacal_griz_run2.1i_dr1b.yaml]: Update path to /projecta path

File: GCRCatalogs/catalog_configs/dc2_metacal_griz_run2.1i_dr1b_tract2897.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+subclass_name: dc2_metacal.DC2MetacalCatalog
+base_dir: /global/homes/f/flanusse/scratch/object_mcal_catalog_coaddv1_dr1b
+schema_filename: schema.yaml
+filename_pattern: 'metacal_tract_2897.parquet$'
+description: DC2 Run 2.1i coadd-v1b Metacal Catalog (griz)
+creators: ['Francois Lanusse', 'Johann Cohen-Tanugi']
+bands: 'griz'
+apply_metacal_test3_fix: true
+
```
Comment (yymao) [inline GCRCatalogs/catalog_configs/dc2_metacal_griz_run2.1i_dr1b_tract2897.yaml]: Better way to write this config is

```yaml
base_on: dc2_metacal_griz_run2.1i_dr1b.yaml
filename_pattern: 'metacal_tract_2897.parquet$'
```

In fact, if this is just used in the composite config, you don't even this standalone config at all. See below.

File: GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_tract2897.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+subclass_name: dc2_object.DC2ObjectCatalog
+base_dir: /global/projecta/projectdirs/lsst/production/DC2_ImSim/Run2.1i/dpdd/calexp-v1:coadd-dr1b-v1/object_table_summary
+schema_filename: trim_schema.yaml
+filename_pattern: 'trim_object_tract_2897\.hdf5$'
+description: DC2 Run 2.1i Object Catalog Tract 2897 (with only DPDD columns and native columns needed for the DPDD columns)
+creators: ['DESC DC2 Team']
+pixel_scale: 0.2
+include_in_default_catalog_list: true
```
Comment (yymao) [inline GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_tract2897.yaml]: No need of this config --- remove. See below.

File: GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_tract2897_with_metacal_griz.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+subclass_name: composite.CompositeReader
+description: DC2 Run 2.1i Object Catalog with metacal (griz)
+catalogs:
+  - catalog_name: dc2_object_run2.1i_dr1b_tract2897
+    matching_method: 'id'
+  - catalog_name: dc2_metacal_griz_run2.1i_dr1b_tract2897
+    matching_method: 'id'
+
```
Comment (yymao) [inline GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_tract2897_with_metacal_griz.yaml]:
```yaml
subclass_name: composite.CompositeReader
description: DC2 Run 2.1i Object Catalog with metacal (griz)
catalogs:
  - catalog_name: dc2_object_run2.1i_dr1b_parquet
    filename_pattern: 'dpdd_object_tract_2897\.parquet$'
    matching_method: MATCHING_FORMAT
  - catalog_name: dc2_metacal_griz_run2.1i_dr1b
    filename_pattern: 'metacal_tract_2897\.parquet$'
    matching_method: MATCHING_FORMAT
```

File: GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_with_metacal_griz.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+subclass_name: composite.CompositeReader
+description: DC2 Run 2.1i Object Catalog with metacal (griz)
+catalogs:
+  - catalog_name: dc2_object_run2.1i_dr1b_all_columns
+    matching_method: 'id'
```
Comment (yymao) [inline GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_with_metacal_griz.yaml]: Change to `MATCHING_FORMAT`

File: GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_with_metacal_griz.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  - catalog_name: dc2_metacal_griz_run2.1i_dr1b
+    matching_method: 'id'
```
Comment (yymao) [inline GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_with_metacal_griz.yaml]: change to `MATCHING_FORMAT`

File: GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_parquet.yaml [hunk 7]
```diff
@@ -1,8 +1,6 @@
-subclass_name: dc2_dm_catalog.DC2DMCatalog
-base_dir: /global/projecta/projectdirs/lsst/production/DC2_ImSim/Run2.1i/dpdd/calexp-v1:coadd-dr1b-v1/object_table_summary
-schema_filename: dpdd_schema.yaml
-filename_pattern: 'dpdd_object_tract_\d+\.parquet$'
-description: DC2 Run 2.1i DPDD Object Table - Identical to dc2_object_run2.1i_dr1b, but with underlying Parquet files instead of HDF5.
+subclass_name: dc2_dm_object.DC2ObjectCatalog
+base_dir: /global/cscratch1/sd/yymao/desc/dc2_object_run2.1i_dr1b
```
Comment (yymao) [inline GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_parquet.yaml]:
```suggestion
base_dir: /global/projecta/projectdirs/lsst/production/DC2_ImSim/Run2.1i/dpdd/calexp-v1:coadd-dr1b-v1/object_table_parquet
```

**Comment by yymao** [comment]
@EiffL can you 
1. rebase this to master,and 
2. check if the record order in the two catalogs are indeed the same? It doesn't seem to be the case for dr1a. If the order is not the same, "MATCHING_FORMAT" or "MATCHING_ORDER" won't work.

File: GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_with_metacal_griz.yaml (hunk 6 above)
Comment (EiffL) [reply GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_with_metacal_griz.yaml]: I'm keeping the matching by ID as it's not significantly slower, much safer, and handles missing patches from metacal

**Comment by EiffL** [comment]
@yymao I was using your improved DMDC2 reader, did you not merge it into master? should I move back to the  version of  dc2_dm_catalog.py in  mamster?

File: GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_tract2897_with_metacal_griz.yaml (hunk 4 above)
Comment (EiffL) [reply GCRCatalogs/catalog_configs/dc2_object_run2.1i_dr1b_tract2897_with_metacal_griz.yaml]: I removed that useless file

**Comment by yymao** [comment]
This is almost ready to merge pending (1) CI test and (2) a confirmation from @EiffL that it works.

**Comment by EiffL** [comment]
It's  all good on my end

**Pull request merged**